### PR TITLE
Changes to actionprn 282464

### PR DESCRIPTION
**Pull request by AtosSK**

### CODE REVIEW

File: EPRN.PRNS.API/Controllers/PRNController.cs [hunk 1]
```diff
@@ -76,7 +76,7 @@ public async Task<IActionResult> GetTonnage(int id)
         public async Task<IActionResult> SaveCheckYourAnswersState(
             int id)
         {
-            await _prnService.SaveCheckYourAnswers(id);
+            await _prnService.SaveCheckYourAnswers(id, "Update");
```
Comment (NikPeters-Atos) [inline EPRN.PRNS.API/Controllers/PRNController.cs:79]: What's this string being passed as a parameter for?

Comment (NikPeters-Atos) [inline EPRN.PRNS.API/Controllers/PRNController.cs]: Firstly, why pass this in from the controller? Set it at the service level - that's what it's for. Secondly, if we are going to be updating status' without a reason, then we _might_ want to consider making reason not a required field. You might want to discuss this.

File: EPRN.PRNS.API/Controllers/PRNController.cs [hunk 2]
```diff
@@ -164,6 +164,32 @@ public async Task<ActionResult> SaveDecemberWaste(int? id, bool decemberWaste)
...
+            
+            if (prnReference == null)
+                return NotFound();
+
+            return Ok(prnReference.ReferenceNumber);
+        }
+
+        [HttpPost]
+        [Route("SentTo/{status}")]
+        public async Task<ActionResult> SaveSentTo(int? id, PrnStatus? status)
```
Comment (NikPeters-Atos) [inline EPRN.PRNS.API/Controllers/PRNController.cs]: Don't make the fields nullable on the API unless that is the actual intension. I suspect not though

File: EPRN.PRNS.API/Repositories/Repository.cs [hunk 3]
```diff
@@ -267,5 +281,14 @@ await _prnContext.PRN
                     sp.SetProperty(wj => wj.DecemberWaste, decemberWaste)
                 );
         }
+
+        public async Task SaveSentTo(int id, string status)
+        {
+            await _prnContext.PRN
+                .Where(wj => wj.Id == id)
+                .ExecuteUpdateAsync(sp =>
+                    sp.SetProperty(wj => wj.SentTo, status)
```
Comment (NikPeters-Atos) [inline EPRN.PRNS.API/Repositories/Repository.cs]: The parameter name doesn't seem to make sense - status? For the SentTo field?

File: EPRN.PRNS.API/Services/Interfaces/IPrnService.cs [hunk 4]
```diff
@@ -25,7 +25,7 @@ public interface IPrnService
 
         Task<StatusAndProducerDto> GetStatusWithProducerName(int id);
 
-        Task SaveCheckYourAnswers(int id);
+        Task SaveCheckYourAnswers(int id, string reason);
```
Comment (NikPeters-Atos) [inline EPRN.PRNS.API/Services/Interfaces/IPrnService.cs]: Remove reason parameter

File: EPRN.PRNS.API/Services/PrnService.cs [hunk 5]
```diff
@@ -136,16 +136,26 @@ public async Task<DecemberWasteDto> GetDecemberWaste(int journeyId)
...
         }
 
-        #region Private methods - Keep at bottom of file
-        // Stub this and generate a random PRN reference number
-        // In time a specific generation algorithm will
-        // be specified
-        private string GenerateReferenceNumber() => $"PRN{Guid.NewGuid().ToString().Replace("-", string.Empty).Substring(0, 10)}";
+        public async Task SaveSentTo(int id, PrnStatus status)
+        {
+            await _prnRepository.SaveSentTo(id, status.ToString());
```
Comment (NikPeters-Atos) [inline EPRN.PRNS.API/Services/PrnService.cs]: Why are we saving the prn status to the sent to field?

File: EPRN.PRNS.API/Services/PrnService.cs [hunk 6]
```diff
@@ -136,16 +136,26 @@ public async Task<DecemberWasteDto> GetDecemberWaste(int journeyId)
...
+        }
+
+    #region Private methods - Keep at bottom of file
+    // Stub this and generate a random PRN reference number
+    // In time a specific generation algorithm will
+    // be specified
+    private string GenerateReferenceNumber() => $"PRN{Guid.NewGuid().ToString().Replace("-", string.Empty).Substring(0, 10)}";
```
Comment (NikPeters-Atos) [inline EPRN.PRNS.API/Services/PrnService.cs]: The indenting has been lost here.

File: EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs [hunk 7]
```diff
@@ -247,6 +247,42 @@ public async Task<IActionResult> DecemberWaste(DecemberWasteViewModel decemberWa
                                     new { decemberWaste.Id });
         }
 
+        [HttpGet]
+        public async Task<IActionResult> ActionPrn(int? id)
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs]: ActionPrn is not very meangingful - if you want to keep that as an action name then use the Action attribute, and then rename the method to something a little bit more meaningful.

File: EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs [hunk 8]
```diff
@@ -108,13 +109,13 @@ public async Task<IActionResult> CheckYourAnswers(CheckYourAnswersViewModel chec
...
-                    area = Category, 
-                    id = checkYourAnswersViewModel.Id 
-                });
+                            Routes.Areas.Actions.PRNS.ActionPrn,
+                            Routes.Areas.Controllers.Reprocessor.PRNS,
+                            new
+                            {
+                                area = Category,
+                                id = checkYourAnswersViewModel.Id
+                            });
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs]: Fix indenting

File: EPRN.Portal/Services/PRNService.cs [hunk 9]
```diff
@@ -224,11 +225,25 @@ public async Task<ViewPRNViewModel> GetViewPrnViewModel(string reference)
...
-            return new ActionPrnViewModel
-            {
-                Id = id
-            };
+            var reference = await _httpPrnsService.GetPrnReference(id);
+            var dto = await _httpPrnsService.GetPrnDetails(reference);
+
+            var actionPrnViewModel = new ActionPrnViewModel { Id = id };
+
+            return actionPrnViewModel;
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Services/PRNService.cs:233]: This is all wrong

File: EPRN.Portal/Services/PRNService.cs [hunk 10]
```diff
@@ -224,11 +225,25 @@ public async Task<ViewPRNViewModel> GetViewPrnViewModel(string reference)
 
         public async Task<ActionPrnViewModel> GetActionPrnViewModel(int id)
         {
-            //TODO: This will need to be populated by the journey so we can return to it.
-            return new ActionPrnViewModel
-            {
-                Id = id
-            };
+            var reference = await _httpPrnsService.GetPrnReference(id);
+            var dto = await _httpPrnsService.GetPrnDetails(reference);
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Services/PRNService.cs]: Why don't you just create your own end point that returns the data that you need instead of making 2 calls? This is very inefficient

File: EPRN.Portal/Services/PRNService.cs [hunk 11]
```diff
@@ -224,11 +225,25 @@ public async Task<ViewPRNViewModel> GetViewPrnViewModel(string reference)
...
+
+            var actionPrnViewModel = new ActionPrnViewModel { Id = id };
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Services/PRNService.cs]: Previous code is not used! Is this unfinished work? If so you need to double check over your work before checking it in. Ensure that it works from all aspects. Have you written unit tests? Because if you have they should be failing, and if they're not, then they are incorrect. If you have not written any, you should do. Every controller and service needs unit tests, even for the most simple bit of code.

File: EPRN.Portal/Views/Areas/Reprocessor/PRNS/ActionPRN.cshtml [hunk 12]
```diff
@@ -0,0 +1,38 @@
+﻿@using EPRN.Common.Enums;
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Views/Areas/Reprocessor/PRNS/ActionPRN.cshtml:1]: This is a PERN, not a PRN - by being in Exporter area makes this a PERN. Therefore any reference to PRN in the view needs to be changed to PERN

File: EPRN.Portal/Views/Areas/Exporter/PRNS/ActionPRN.cshtml
Comment (NikPeters-Atos) [inline EPRN.Portal/Views/Areas/Exporter/PRNS/ActionPRN.cshtml:1]: This is a PERN, not a PRN - by being in Exporter area makes this a PERN. Therefore any reference to PRN in the view needs to be changed to PERN

File: EPRN.Portal/Views/Areas/Exporter/PRNS/PrnSavedAsDraftConfirmation.cshtml
Comment (NikPeters-Atos) [inline EPRN.Portal/Views/Areas/Exporter/PRNS/PrnSavedAsDraftConfirmation.cshtml:1]: This again needs to refer to PERN not PRN

File: EPRN.UnitTests/Portal/Controllers/PrnsControllerTests.cs [hunk 13]
```diff
@@ -21,85 +21,6 @@ public void Init()
             _prnsController = new PrnsController(_mockPrnService.Object);
         }
 
-        [TestMethod]
```
Comment (NikPeters-Atos) [inline EPRN.UnitTests/Portal/Controllers/PrnsControllerTests.cs:24]: Why have you deleted a load of unit tests?

File: EPRN.PRNS.API/Controllers/PRNController.cs (hunk 1 above)
Comment (AtosSK) [reply EPRN.PRNS.API/Controllers/PRNController.cs]: Done.

File: EPRN.PRNS.API/Controllers/PRNController.cs (hunk 2 above)
Comment (AtosSK) [reply EPRN.PRNS.API/Controllers/PRNController.cs]: Done.

File: EPRN.PRNS.API/Repositories/Repository.cs (hunk 3 above)
Comment (AtosSK) [reply EPRN.PRNS.API/Repositories/Repository.cs]: Done.

File: EPRN.PRNS.API/Services/Interfaces/IPrnService.cs (hunk 4 above)
Comment (AtosSK) [reply EPRN.PRNS.API/Services/Interfaces/IPrnService.cs]: Done.

File: EPRN.PRNS.API/Services/PrnService.cs (hunk 5 above)
Comment (AtosSK) [reply EPRN.PRNS.API/Services/PrnService.cs]: Done.

File: EPRN.PRNS.API/Services/PrnService.cs (hunk 6 above)
Comment (AtosSK) [reply EPRN.PRNS.API/Services/PrnService.cs]: Done.

File: EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs (hunk 7 above)
Comment (AtosSK) [reply EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs]: Done.

File: EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs (hunk 8 above)
Comment (AtosSK) [reply EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs]: Done.

File: EPRN.Portal/Services/PRNService.cs (hunk 10 above)
Comment (AtosSK) [reply EPRN.Portal/Services/PRNService.cs]: Done.

File: EPRN.Portal/Views/Areas/Exporter/PRNS/ActionPRN.cshtml
Comment (AtosSK) [reply EPRN.Portal/Views/Areas/Exporter/PRNS/ActionPRN.cshtml:1]: Done.

File: EPRN.Portal/Views/Areas/Exporter/PRNS/PrnSavedAsDraftConfirmation.cshtml
Comment (AtosSK) [reply EPRN.Portal/Views/Areas/Exporter/PRNS/PrnSavedAsDraftConfirmation.cshtml:1]: Done.

File: EPRN.UnitTests/Portal/Controllers/PrnsControllerTests.cs (hunk 13 above)
Comment (AtosSK) [reply EPRN.UnitTests/Portal/Controllers/PrnsControllerTests.cs:24]: I moved these tests.

Comment (AtosSK) [reply EPRN.UnitTests/Portal/Controllers/PrnsControllerTests.cs:24]: Added some more tests.

File: EPRN.Portal/Services/PRNService.cs (hunk 11 above)
Comment (AtosSK) [reply EPRN.Portal/Services/PRNService.cs]: Done.

File: EPRN.Common/Constants/Constants.cs [hunk 14]
```diff
@@ -69,6 +69,8 @@ public static class PRNS
                     public const string Cancelled = "Cancelled";
                     public const string RequestCancelConfirmed = "RequestCancelConfirmed";
                     public const string DecemberWaste = "DecemberWaste";
+                    public const string DestinationPrn = "DestinationPrn";
+                    public const string PrnSavedAsDraftConfirmation = "PrnSavedAsDraftConfirmation";
```
Comment (NikPeters-Atos) [inline EPRN.Common/Constants/Constants.cs:79]: I would have a snappier action name if it were me: DraftConfirmation. I don't think there's any need for it to be any longer than that. It's already in a controller called PRN, so we don;t need that. The rest is implied.

File: EPRN.PRNS.API/Controllers/PRNController.cs [hunk 15]
```diff
@@ -164,6 +164,44 @@ public async Task<ActionResult> SaveDecemberWaste(int? id, bool decemberWaste)
             return Ok();
         }
 
+        [HttpGet]
+        [Route("Reference")]
+        public async Task<IActionResult> GetPrnReference(int id)
+        {
+            var prnDetailsDto = await _prnService.GetPrnDetails(id);
+            
+            if (prnDetailsDto == null)
```
Comment (NikPeters-Atos) [inline EPRN.PRNS.API/Controllers/PRNController.cs]: I wouldn't worry about that - the API has a facility for checking that the record exists. If the dto comes over as null, then we have different issues to it not being found

File: EPRN.PRNS.API/Services/PrnService.cs [hunk 16]
```diff
@@ -74,9 +74,10 @@ public async Task<StatusAndProducerDto> GetStatusWithProducerName(int id)
 
         public async Task SaveCheckYourAnswers(int id)
         {
+            string reason = "Update";
             await _prnRepository.UpdatePrnStatus(
                 id, 
-                PrnStatus.CheckYourAnswersComplete);
+                PrnStatus.CheckYourAnswersComplete, reason);
```
Comment (NikPeters-Atos) [inline EPRN.PRNS.API/Services/PrnService.cs]: I think we should remove the Required attribute from the history reason field. If we are adding history where we do not need a reason, there's no point in creating random text to satisfy the not null behaviour. It just means it should be a nullable field.

File: EPRN.PRNS.API/Services/PrnService.cs [hunk 17]
```diff
@@ -136,16 +137,34 @@ public async Task<DecemberWasteDto> GetDecemberWaste(int journeyId)
...
+
         #region Private methods - Keep at bottom of file
         // Stub this and generate a random PRN reference number
         // In time a specific generation algorithm will
         // be specified
         private string GenerateReferenceNumber() => $"PRN{Guid.NewGuid().ToString().Replace("-", string.Empty).Substring(0, 10)}";
-        #endregion
-    }
+            #endregion
+        }
```
Comment (NikPeters-Atos) [inline EPRN.PRNS.API/Services/PrnService.cs:172]: Indenting looks a little wrong here

File: EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs [hunk 18]
```diff
@@ -107,13 +108,11 @@ public async Task<IActionResult> CheckYourAnswers(CheckYourAnswersViewModel chec
...
             await _prnService.SaveCheckYourAnswers(checkYourAnswersViewModel.Id);
 
-            return RedirectToAction(
-                Routes.Areas.Actions.PRNS.WhatToDo, 
-                Routes.Areas.Controllers.Reprocessor.PRNS, 
-                new 
-                { 
-                    area = Category, 
-                    id = checkYourAnswersViewModel.Id 
+            return RedirectToAction(Routes.Areas.Actions.PRNS.DestinationPrn, Routes.Areas.Controllers.Reprocessor.PRNS,
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs]: When parameters go on a long list, move each one on to a new line

File: EPRN.Portal/Services/PRNService.cs [hunk 19]
```diff
@@ -221,13 +222,28 @@ public async Task<ViewPRNViewModel> GetViewPrnViewModel(string reference)
...
+
+        public async Task SaveDraftPrn(DestinationPrnViewModel destinationPrnViewModel)
+        {
+            if (destinationPrnViewModel == null)
+                throw new ArgumentNullException(nameof(destinationPrnViewModel));
+
+            if (destinationPrnViewModel.DoWithPRN == null)
+                throw new ArgumentNullException(nameof(destinationPrnViewModel.DoWithPRN));
+
+            destinationPrnViewModel.Reason = "Update";
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Services/PRNService.cs]: Yeah, let's get rid of reason being a required field.

File: EPRN.UnitTests/API/PRNS/Services/PrnServiceTests.cs [hunk 20]
```diff
@@ -124,16 +124,16 @@ public async Task SaveCheckYourAnswers_CallsService_WithComplete()
...
             // act
             await _prnService.SaveCheckYourAnswers(id);
 
             // assert
             _mockRepository.Verify(s =>
                 s.UpdatePrnStatus(
-                    It.Is<int>(p => p == id),
-                    It.Is<PrnStatus>(p => p == PrnStatus.CheckYourAnswersComplete),
-                    It.Is<string>(p => p == null)),
+                    It.IsAny<int>(),
```
Comment (NikPeters-Atos) [inline EPRN.UnitTests/API/PRNS/Services/PrnServiceTests.cs]: Since you're verifying a call has been made, you should know the parameters. I'm not sure why you removed all the It.Is for It.IsAny.

File: EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs [hunk 21]
```diff
@@ -231,6 +226,43 @@ public async Task<IActionResult> DecemberWaste(DecemberWasteViewModel decemberWa
...
+        public async Task<IActionResult> DestinationPrn(DestinationPrnViewModel destinationPrnViewModel)
+        {
+            if (!ModelState.IsValid)
+                return View(destinationPrnViewModel);
+
+            if (destinationPrnViewModel.DoWithPRN == PrnStatus.Draft)
+            {
+                await _prnService.SaveDraftPrn(destinationPrnViewModel);
+                return RedirectToAction(Routes.Areas.Actions.PRNS.PrnSavedAsDraftConfirmation,
+                                new { area = Category.ToString(), destinationPrnViewModel.Id });
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs]: Can we just go through this when you are available? I want to check the operation of this bit.

File: EPRN.Common/Constants/Constants.cs (hunk 14 above)
Comment (AtosSK) [reply EPRN.Common/Constants/Constants.cs:79]: Done.

File: EPRN.PRNS.API/Services/PrnService.cs (hunk 17 above)
Comment (AtosSK) [reply EPRN.PRNS.API/Services/PrnService.cs:172]: Done.

File: EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs (hunk 18 above)
Comment (AtosSK) [reply EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs]: Done.

File: EPRN.PRNS.API/Services/PrnService.cs (hunk 16 above)
Comment (AtosSK) [reply EPRN.PRNS.API/Services/PrnService.cs]: Done.

File: EPRN.Portal/Services/PRNService.cs (hunk 19 above)
Comment (AtosSK) [reply EPRN.Portal/Services/PRNService.cs]: Done.

File: EPRN.UnitTests/API/PRNS/Services/PrnServiceTests.cs (hunk 20 above)
Comment (AtosSK) [reply EPRN.UnitTests/API/PRNS/Services/PrnServiceTests.cs]: Done.

File: EPRN.PRNS.API/Controllers/PRNController.cs (hunk 15 above)
Comment (AtosSK) [reply EPRN.PRNS.API/Controllers/PRNController.cs]: Done.

File: EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs (hunk 21 above)
Comment (AtosSK) [reply EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs]: Done.

File: EPRN.Common.Data/DataModels/PrnHistory.cs [hunk 22]
```diff
@@ -12,8 +13,8 @@ public class PrnHistory : IdBaseEntity
         public PrnStatus Status { get; set; }
 
         [MaxLength(200)]
-        [Required]
-        public string Reason { get; set; }
+        [AllowNull]
+        public string? Reason { get; set; }
```
Comment (NikPeters-Atos) [inline EPRN.Common.Data/DataModels/PrnHistory.cs:15]: Don't think this is required, is it? Strings are nullable, and therefore just a regular string will denote a field that will allow nulls.

File: EPRN.Common.Data/Migrations/20240214120931_ReasonNowNotRequired.cs [hunk 23]
```diff
@@ -0,0 +1,40 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
```
Comment (NikPeters-Atos) [inline EPRN.Common.Data/Migrations/20240214120931_ReasonNowNotRequired.cs]: When you see to my comment above, delete this migration and re-generate

File: EPRN.Common.Data/Migrations/20240214124038_Reason-Allow-Null.cs [hunk 24]
```diff
@@ -0,0 +1,22 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
```
Comment (NikPeters-Atos) [inline EPRN.Common.Data/Migrations/20240214124038_Reason-Allow-Null.cs]: You're not checking you PRs again - why is this here?

File: EPRN.Common.Data/Migrations/20240214124135_Reason-Allow-Null-Field.cs [hunk 25]
```diff
@@ -0,0 +1,22 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
```
Comment (NikPeters-Atos) [inline EPRN.Common.Data/Migrations/20240214124135_Reason-Allow-Null-Field.cs]: Check PRs - another migration that should have been noted, and removed before submitting for review

File: EPRN.PRNS.API/Controllers/PRNController.cs [hunk 26]
```diff
@@ -164,6 +164,41 @@ public async Task<ActionResult> SaveDecemberWaste(int? id, bool decemberWaste)
...
+        [Route("Details")]
+        public async Task<IActionResult> GetPrnDetails(int id)
+        {
+            var prnDetailsDto = await _prnService.GetPrnDetails(id);
+            
+            return Ok(prnDetailsDto);
+        }
+
+        [HttpPost]
+        [Route("SentTo/{sentTo}")]
```
Comment (NikPeters-Atos) [inline EPRN.PRNS.API/Controllers/PRNController.cs]: The value of SentTo should come from the body of the request, not the url. This is because it's a string, and could be almost any length. URLs have a limit of the length, but body content doesn't

File: EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs [hunk 27]
```diff
@@ -109,12 +110,11 @@ public async Task<IActionResult> CheckYourAnswers(CheckYourAnswersViewModel chec
             await _prnService.SaveCheckYourAnswers(checkYourAnswersViewModel.Id);
 
             return RedirectToAction(
-                Routes.Areas.Actions.PRNS.WhatToDo,
-                Routes.Areas.Controllers.Exporter.PRNS, 
-                new 
-                { 
+                Routes.Areas.Actions.PRNS.DraftConfirmation,
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs:111]: Indenting seems wrong here

File: EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs [hunk 28]
```diff
@@ -231,6 +233,33 @@ public async Task<IActionResult> DecemberWaste(DecemberWasteViewModel decemberWa
...
 
+        [HttpGet]
+        [ActionName(Routes.Areas.Actions.PRNS.DraftConfirmation)]
+        public async Task<IActionResult> DraftConfirmation(int? id)
+        {
+            if (id == null)
+                return NotFound();
+
+            var viewModel = await _prnService.GetDraftConfirmationViewModel(id.Value);
+            return View(viewModel);
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs:249]: What happens if you land on this page when the PRN has already been created as a draft? I thought you were going to see to this

File: EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs [hunk 29]
```diff
@@ -107,13 +108,11 @@ public async Task<IActionResult> CheckYourAnswers(CheckYourAnswersViewModel chec
...
             await _prnService.SaveCheckYourAnswers(checkYourAnswersViewModel.Id);
 
-            return RedirectToAction(
-                Routes.Areas.Actions.PRNS.WhatToDo, 
-                Routes.Areas.Controllers.Reprocessor.PRNS, 
-                new 
-                { 
-                    area = Category, 
-                    id = checkYourAnswersViewModel.Id 
+            return RedirectToAction(Routes.Areas.Actions.PRNS.DraftConfirmation, Routes.Areas.Controllers.Reprocessor.PRNS,
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs:110]: Why have you edited this code, and made the parameters worse?

File: EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs [hunk 30]
```diff
@@ -236,6 +235,33 @@ public async Task<IActionResult> DecemberWaste(DecemberWasteViewModel decemberWa
...
         }
 
+        [HttpGet]
+        [ActionName(Routes.Areas.Actions.PRNS.DraftConfirmation)]
+        public async Task<IActionResult> DraftConfirmation(int? id)
+        {
+            if (id == null)
+                return NotFound();
+
+            var viewModel = await _prnService.GetDraftConfirmationViewModel(id.Value);
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs:244]: Again, what happens if you land on this page with a PRN that is already a draft?

File: EPRN.Portal/Profiles/PRNProfile.cs [hunk 31]
```diff
@@ -31,6 +31,13 @@ public PRNProfile()
 
             CreateMap<PRNHistoryDto, PRNHistoryViewModel>();
             CreateMap<StatusAndProducerDto, CancelViewModel>();
+            CreateMap<PRNDetailsDto, DraftConfirmationViewModel>()
+                .ForMember(d => d.Reference, o => o.MapFrom(s => s.ReferenceNumber))
+                .ForMember(d => d.DoWithPRN, o => o.MapFrom(s =>
```
Comment (NikPeters-Atos) [inline EPRN.Portal/Profiles/PRNProfile.cs]: Shouldn't the most recent status be retrieved at the API Iside, rather than retrieve all of them and filter them out at the Portal side? Seems inefficient to me

File: EPRN.Common.Data/DataModels/PrnHistory.cs (hunk 22 above)
Comment (AtosSK) [reply EPRN.Common.Data/DataModels/PrnHistory.cs:15]: Done.

File: EPRN.Common.Data/Migrations/20240214124038_Reason-Allow-Null.cs (hunk 24 above)
Comment (AtosSK) [reply EPRN.Common.Data/Migrations/20240214124038_Reason-Allow-Null.cs]: Done.

File: EPRN.PRNS.API/Controllers/PRNController.cs (hunk 26 above)
Comment (AtosSK) [reply EPRN.PRNS.API/Controllers/PRNController.cs]: Done.

File: EPRN.Common.Data/Migrations/20240214124135_Reason-Allow-Null-Field.cs (hunk 25 above)
Comment (AtosSK) [reply EPRN.Common.Data/Migrations/20240214124135_Reason-Allow-Null-Field.cs]: Deleted.

File: EPRN.Common.Data/Migrations/20240214120931_ReasonNowNotRequired.cs (hunk 23 above)
Comment (AtosSK) [reply EPRN.Common.Data/Migrations/20240214120931_ReasonNowNotRequired.cs]: Done.

File: EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs (hunk 27 above)
Comment (AtosSK) [reply EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs:111]: Done.

File: EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs (hunk 29 above)
Comment (AtosSK) [reply EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs:110]: Done.

File: EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs (hunk 28 above)
Comment (AtosSK) [reply EPRN.Portal/Areas/Exporter/Controllers/PRNSController.cs:249]: Done.

File: EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs (hunk 30 above)
Comment (AtosSK) [reply EPRN.Portal/Areas/Reprocessor/Controllers/PRNSController.cs:244]: Done.

File: EPRN.Portal/Profiles/PRNProfile.cs (hunk 31 above)
Comment (AtosSK) [reply EPRN.Portal/Profiles/PRNProfile.cs]: Done.

**Review by NikPeters-Atos — APPROVED** [review]
